### DISCUSSION TITLE
Question: is it possible to perform static analysis on queries to detect errors? (= a linter)

### DISCUSSION THREAD

**Discussion by Lucas-C** [trinodb/trino#23988] [Q&A]
Hi,

I am looking for a tool to perform **static query analysis**,
and be able to detect errors like `XXX must be an aggregate expression or appear in GROUP BY clause`
without the need to deploy `trino` nor have a database with a vali schema.

It does not seem that [`trino` CLI](https://trino.io/docs/current/client/cli.html) offers a query linter.

External tools like [sqlfluff](https://github.com/sqlfluff/sqlfluff) are not able to detect errors like `XXX must be an aggregate expression or appear in GROUP BY clause`:
```
$ sqlfluff lint query-with-error.zoql --dialect trino
All Finished 📜 🎉!
```

I think I found that this specific error is raised by [AggregationAnalyzer.java](https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/sql/analyzer/AggregationAnalyzer.java).

Hence my question: **is it possible to invoke some Java class/method to perform just query analysis** on some `sql` query string input?

PS: The link to your Slack channel seems dead there: https://trino.io/slack

**Comment by nineinchnick** [comment]
The Trino parser is not a stable API, so using it for this purpose is not officially supported, but it was raised before:
* https://github.com/trinodb/trino/pull/13044
* https://github.com/trinodb/trino/pull/19433
* https://github.com/kokosing/trino-query-formatter

Obviously, only formatting a query is not the same as analyzing it, but it can also catch some errors.

**Comment by Lucas-C** [comment]
Thank you very much for your answer @nineinchnick 

I tested https://github.com/kokosing/trino-query-formatter but it does not detect the issue in my query:
```
$ ./mvnw install -DskipTests=true
$ java -jar target/trino-query-formatter-*-executable.jar < query-with-error.zoql
```
It seems that it does not call the `io.trino.sql.analyzer` package.

PR https://github.com/trinodb/trino/pull/13044 seemed very interesting.
It looks like it tried to incorporate https://github.com/kokosing/trino-query-formatter into `trinodb/trino`
Out of curiosity, would a `--format` argument be a welcome addition to the `trino` CLI?

I'm a bit sad that both PRs were closed, but they are interesting starting points 👍

**Comment by nineinchnick** [comment]
The issues/PRs are closed, because the ideas were rejected.
